### PR TITLE
Jit bind

### PR DESCRIPTION
**Pull request by Hardcode84**
bind-like functionality which allow to generate efficient specialized versions of functions (much like easy::jit).

Add `ldc.dynamic_compile.bind` function with interface similar to C++ `std::bind`. Function returns `BindPtr` object (which is reference counted internally) with `opCall` method.

Actual body is generated during `compileDynamicCode` and after that `BindPtr` can be called.
Function passed to `bind` must be marked `@dynamicCompile`, otherwise `BindPtr.isCallable()` will return `false` and 'opCall' asserts.

If function bind as parameter marked as `@dynamicCompile` it will be efficiently optimized (see `dynamiccompile/bind_func_opt.d` test).

Also bind functions can be chained together (see `dynamiccompile/bind_nested_opt.d` test).

### CODE REVIEW

**Comment by Hardcode84** [comment]
Ready, I think.

**Comment by JohanEngelen** [comment]
Great that you submitted some parts as separate PRs.
Quickly glancing over it, I missed documentation of the new user function `bind`.

File: runtime/jit-rt/d/ldc/dynamic_compile.d [hunk 1]
```diff
@@ -83,10 +83,306 @@ void compileDynamicCode(in CompilerSettings settings = CompilerSettings.init)
   rtCompileProcessImpl(context, context.sizeof);
 }
 
+auto bind(F, Args...)(F func, Args args)
```
Comment (JohanEngelen) [inline runtime/jit-rt/d/ldc/dynamic_compile.d]: Add docs

File: runtime/jit-rt/d/ldc/dynamic_compile.d [hunk 2]
```diff
@@ -83,10 +83,306 @@ void compileDynamicCode(in CompilerSettings settings = CompilerSettings.init)
...
+{
+  import std.format;
+  static assert(isFunctionPointer!F);
```
Comment (JohanEngelen) [inline runtime/jit-rt/d/ldc/dynamic_compile.d]: nice to add a message to the user when this fails

File: runtime/jit-rt/d/ldc/dynamic_compile.d [hunk 3]
```diff
@@ -83,10 +83,306 @@ void compileDynamicCode(in CompilerSettings settings = CompilerSettings.init)
...
+    alias UnbindTypes = UnbindTypes!(Index[1..$], Args[1..$]);
+  }
+}
+
+struct BindPayloadBase(F)
+{
+  alias FuncParams = Parameters!(F);
+  alias Ret = ReturnType!F;
+  F func = null;
+  static assert(func.offsetof == 0, "func must be fist");
```
Comment (JohanEngelen) [inline runtime/jit-rt/d/ldc/dynamic_compile.d]: typo in error msg

**Comment by Hardcode84** [comment]
just rebase (and something bad happened with commit dates so github display nonsense like ` 8 days ago`)

File: runtime/jit-rt/d/ldc/dynamic_compile.d (hunk 1 above)
Comment (Hardcode84) [reply runtime/jit-rt/d/ldc/dynamic_compile.d]: done

File: runtime/jit-rt/d/ldc/dynamic_compile.d (hunk 2 above)
Comment (Hardcode84) [reply runtime/jit-rt/d/ldc/dynamic_compile.d]: done

**Comment by Hardcode84** [comment]
rebased, added docs

File: runtime/jit-rt/d/ldc/dynamic_compile.d [hunk 4]
```diff
@@ -83,10 +83,335 @@ void compileDynamicCode(in CompilerSettings settings = CompilerSettings.init)
...
+  enum ParametersCount = FuncParams.length;
+  static assert(ParametersCount == Args.length, format("Invalid bind parameter count: %s, expected %s", Args.length, ParametersCount));
+  assert(func !is null);
+  enum Index = bindParamsInd!(0, 0, Args)();
+  alias PartialF = ReturnType!F function(UnbindTypes!(Index, FuncParams));
+  alias BindPtrType = BindPtr!PartialF;
+  return BindPtrType.make!Index(func, mapBindParams!(F, 0)(args).expand);
+}
+
+immutable placeholder = _placeholder();
```
Comment (JohanEngelen) [inline runtime/jit-rt/d/ldc/dynamic_compile.d]: This member should also be documented briefly. Probably just a reference to `bind`.

**Comment by Hardcode84** [comment]
Updated: delegate as first parameter, `@dynamicCompileEmit`, some docs

File: runtime/jit-rt/d/ldc/dynamic_compile.d (hunk 4 above)
Comment (Hardcode84) [reply runtime/jit-rt/d/ldc/dynamic_compile.d]: done

**Comment by Hardcode84** [comment]
Rebased.

TODO: probably need to write some design overview

**Comment by Hardcode84** [comment]
ready

File: runtime/jit-rt/cpp-so/bind.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#ifndef BIND_H
```
Comment (JohanEngelen) [inline runtime/jit-rt/cpp-so/bind.h]: missing header

File: runtime/jit-rt/cpp-so/param_slice.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
+//===-- param_slice.h - jit support -----------------------------*- C++ -*-===//
+//
+//                         LDC – the LLVM D compiler
+//
+// This file is distributed under the Boost Software License. See the LICENSE
+// file for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// ParamSlice declaration.
```
Comment (JohanEngelen) [inline runtime/jit-rt/cpp-so/param_slice.h]: ;-)   nitpick: this comment is useless, I can read the same a few lines down in C++. What would help is to explain what ParamSlice is used for.

File: docs/dynamic_compile_bind.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
+# Dynamic compile bind
```
Comment (JohanEngelen) [inline docs/dynamic_compile_bind.md]: I like that you added some design documentation

File: CMakeLists.txt [hunk 8]
```diff
@@ -405,7 +405,7 @@ if(LDC_DYNAMIC_COMPILE STREQUAL "AUTO")
     if (NOT (LDC_LLVM_VER LESS 500))
         set(LDC_DYNAMIC_COMPILE True)
         add_definitions(-DLDC_DYNAMIC_COMPILE)
-        add_definitions(-DLDC_DYNAMIC_COMPILE_API_VERSION=1)
+        add_definitions(-DLDC_DYNAMIC_COMPILE_API_VERSION=2)
```
Comment (JohanEngelen) [inline CMakeLists.txt]: Make sure to add an item about this to the release notes.

Comment (JohanEngelen) [inline CMakeLists.txt]: Make sure to add an item about this to the release notes.

**Comment by JohanEngelen** [comment]
I've made some grammar fixes to the documentation.
Overall, I think it's fine to include, I didn't do a review of the actual tech content of this PR. I trust you know what you are doing :)

**Comment by JohanEngelen** [comment]
I've made some grammar fixes to the documentation.
Overall, I think it's fine to include, I didn't do a review of the actual tech content of this PR. I trust you know what you are doing :)

File: runtime/jit-rt/cpp-so/bind.h (hunk 5 above)
Comment (Hardcode84) [reply runtime/jit-rt/cpp-so/bind.h]: fixed

File: runtime/jit-rt/cpp-so/param_slice.h (hunk 6 above)
Comment (Hardcode84) [reply runtime/jit-rt/cpp-so/param_slice.h]: updated

**Pull request merged**